ninshiki_yolo/viewer.py

```diff
@@ -19,41 +19,93 @@
 # THE SOFTWARE.
 
 import argparse
+import cv2
+import numpy as np
 import rclpy
 from rclpy.node import MsgType
 from rclpy.node import Node
-from ninshiki_interfaces.msg import YoloDetectedObjects
+from ninshiki_interfaces.msg import DetectedObjects
+from shisen_interfaces.msg import Image
 
 
 class Viewer(Node):
-    def __init__(self, node_name: str, topic_name: str):
+    def __init__(self, node_name: str, detection_topic: str, img_topic: str):
         super().__init__(node_name)
+        self.detection_result = DetectedObjects()
+
+        self.detected_object_subscription = self.create_subscription(
+            DetectedObjects, detection_topic, self.listener_callback_msg, 10)
+        self.get_logger().info("subscribe detection result on "
+                               + self.detected_object_subscription.topic_name)
 
         self.image_subscription = self.create_subscription(
-            YoloDetectedObjects, topic_name, self.listener_callback, 10)
+            Image, img_topic, self.listener_callback_img, 10)
         self.get_logger().info("subscribe image on " + self.image_subscription.topic_name)
 
-    def listener_callback(self, message: MsgType):
-        for detected_object in message.detected_objects:
-            print(detected_object)
-        print()
+    def listener_callback_msg(self, message: MsgType):
+        self.detection_result = message
+
+    def listener_callback_img(self, message: MsgType):
+        received_frame = np.array(message.data)
+        received_frame = np.frombuffer(received_frame, dtype=np.uint8)
+
+        # Raw Image
+        if (message.quality < 0):
+            received_frame = received_frame.reshape(message.rows, message.cols, 3)
+        # Compressed Image
+        else:
+            received_frame = cv2.imdecode(received_frame, cv2.IMREAD_UNCHANGED)
+
+        if (received_frame.size != 0):
+            detection_frame = self.draw_detection_result(received_frame, self.detection_result)
+            cv2.imshow(self.image_subscription.topic_name, detection_frame)
+            cv2.waitKey(1)
+            self.get_logger().debug("once, received image and display it")
+        else:
+            self.get_logger().warn("once, received empty image")
+
+    def draw_detection_result(self, frame: np.array, detection_result: MsgType) -> np.ndarray:
```

Comment:
As it's in a Python module, I think it's better to put this function into a separate file, to be used by both viewer and detector.